### PR TITLE
US041 - Authorisation System

### PR DESCRIPTION
**Pull request by LMBishop**
Corresponding frontend PR: https://github.com/anthony-raybould/bham1-front-end/pull/14

This PR adds an authorisation system using Dropwizard Auth Filters, and expects user tokens to be passed in the Authorization header. It also adds a new endpoint `/api/whoami` to get a user from a token.

### CODE REVIEW

File: src/main/java/org/kainos/ea/resources/AuthController.java [hunk 1]
```diff
@@ -40,4 +58,15 @@ public Response login(Login login) {
...
+
+    @GET
+    @Path("/whoami")
+    @PermitAll
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation(value = "Returns the user that is currently logged in", response = UserResponse.class, authorizations = {
+        @Authorization(value = HttpHeaders.AUTHORIZATION)
+    })
+    public Response whoami(@Auth @ApiParam(hidden = true) User user) {
+        return Response.ok().entity(new UserResponse(user)).build();
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/resources/AuthController.java:70]: Missing error handling in the endpoint

File: src/main/java/org/kainos/ea/cli/Role.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package org.kainos.ea.cli;
+
+public enum Role {
+
+    ADMIN(1, "Admin"),
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/cli/Role.java]: How are we handling roles, do we have a DB table of the roles or are we using a magic number

Comment (LMBishop) [reply src/main/java/org/kainos/ea/cli/Role.java]: Replaced with https://github.com/anthony-raybould/bham1-back-end/pull/9/files#diff-d61e1278ed1c721e92809a0b4eece6c4b531f163a915224317e36e64eda034c8

File: src/main/java/org/kainos/ea/cli/Role.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        return roleName;
+    }
+
+    @Override
+    public String toString() {
+        return roleName;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/cli/Role.java:30]: As there is a bit of logic here, you should really unit-test it.

File: src/main/java/org/kainos/ea/db/AuthDao.java [hunk 4]
```diff
@@ -43,4 +42,29 @@ public int getUserId(String email) throws SQLException {
         }
         return -1;
     }
+
+    public User getUser(int userId) throws SQLException {
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/db/AuthDao.java:46]: Think you are missing unit tests for this method

File: src/test/java/org/kainos/ea/dao/AuthDaoTest.java [hunk 5]
```diff
@@ -83,4 +84,28 @@ public void getUserId_shouldReturnUserId_whenValid() throws SQLException {
         int result = authDao.getUserId("email@email.com");
         assertEquals(result, 1);
     }
+
+    @Test
```
Comment (TomBillingtonUK) [inline src/test/java/org/kainos/ea/dao/AuthDaoTest.java:87]: Missing a test for the return null path for the getUser method

**Review by TomBillingtonUK — APPROVED** [review]

**Review by RobertCrabtreeKainos — APPROVED** [review]
leo pr so lgtm
